test/TritonGPU/hoist-tmem-alloc.mlir

```diff
@@ -307,3 +307,63 @@ module attributes {"ttg.num-ctas" = 1 : i32, "ttg.num-warps" = 4 : i32, ttg.targ
+  // CHECK-LABEL: @matmul_kernel_tma_persistent_nested
+  tt.func public @matmul_kernel_tma_persistent_nested(%arg0: !tt.tensordesc<tensor<128x32xf8E4M3FN, #shared>>, %arg1: i32, %arg2: i32, %arg3: i64, %arg4: i64, %arg5: !tt.tensordesc<tensor<128x32xf8E4M3FN, #shared>>, %arg6: i32, %arg7: i32, %arg8: i64, %arg9: i64, %arg10: !tt.tensordesc<tensor<128x64xf8E4M3FN, #shared1>>, %arg11: i32, %arg12: i32, %arg13: i64, %arg14: i64, %arg15: i32 {tt.divisibility = 16 : i32}, %arg16: i32 {tt.divisibility = 16 : i32}, %arg17: i32 {tt.divisibility = 16 : i32}) attributes {noinline = false} {
+    %false = arith.constant false
+    %true = arith.constant true
+    // CHECK: %[[ZERO:.*]] = arith.constant 0 : i32
+    %c0_i32 = arith.constant 0 : i32
+    %c1_i32 = arith.constant 1 : i32
+    %c4_i32 = arith.constant 4 : i32
+    %c32_i32 = arith.constant 32 : i32
+    %cst = arith.constant dense<0.000000e+00> : tensor<128x128xf32, #blocked>
+    scf.for %arg18 = %c0_i32 to %c4_i32 step %c1_i32 : i32 {
+      // CHECK: ttng.tmem_alloc
+      // CHECK-NEXT: scf.for %[[ITER:.*]] = %[[START:.*]] to %[[LIMIT:.*]] step
+      %20:3 = scf.for %arg19 = %arg11 to %arg12 step %c1_i32 iter_args(%arg20 = %cst, %arg21 = %c0_i32, %arg22 = %false) -> (tensor<128x128xf32, #blocked>, i32, i1)  : i32 {
+        %28 = tt.descriptor_load %arg0[%arg19, %arg21] : !tt.tensordesc<tensor<128x32xf8E4M3FN, #shared>> -> tensor<128x32xf8E4M3FN, #blocked1>
+        %29 = ttg.local_alloc %28 : (tensor<128x32xf8E4M3FN, #blocked1>) -> !ttg.memdesc<128x32xf8E4M3FN, #shared, #smem>
+        %30 = tt.descriptor_load %arg5[%arg19, %arg21] : !tt.tensordesc<tensor<128x32xf8E4M3FN, #shared>> -> tensor<128x32xf8E4M3FN, #blocked1>
+        %31 = ttg.local_alloc %30 : (tensor<128x32xf8E4M3FN, #blocked1>) -> !ttg.memdesc<128x32xf8E4M3FN, #shared, #smem>
+        %32 = ttg.memdesc_trans %31 {order = array<i32: 1, 0>} : !ttg.memdesc<128x32xf8E4M3FN, #shared, #smem> -> !ttg.memdesc<32x128xf8E4M3FN, #shared2, #smem>
+        %acc, %acc_tok = ttng.tmem_alloc %arg20 : (tensor<128x128xf32, #blocked>) -> (!ttg.memdesc<128x128xf32, #tmem, #ttng.tensor_memory, mutable>, !ttg.async.token)
+        %mma_tok = ttng.tc_gen5_mma %29, %32, %acc[%acc_tok], %arg22, %true : !ttg.memdesc<128x32xf8E4M3FN, #shared, #smem>, !ttg.memdesc<32x128xf8E4M3FN, #shared2, #smem>, !ttg.memdesc<128x128xf32, #tmem, #ttng.tensor_memory, mutable>
+        %34, %load_tok = ttng.tmem_load %acc[%mma_tok] : !ttg.memdesc<128x128xf32, #tmem, #ttng.tensor_memory, mutable> -> tensor<128x128xf32, #blocked>
+        %35 = arith.addi %arg21, %c32_i32 : i32
+        scf.yield %34, %35, %true : tensor<128x128xf32, #blocked>, i32, i1
+      }
+      // CHECK: %[[DIFF:.*]] = arith.subi %[[LIMIT]], %[[START]]
+      // CHECK: %[[COND:.*]] = arith.cmpi sle, %[[DIFF]], %[[ZERO]]
+      // CHECK: %[[TOK:.*]] = scf.if %[[COND]]
+      // CHECK-NEXT:   %[[NTOK:.*]] = ttng.tmem_store
```

Comment:
why not use a predicated store?

Reply:
+1 predicate would be better as control flow tends to block further transformations

Reply:
I disagree, I think predicates make the IR harder to reason about:
  If you have an operation that is predicated, and you need to make changes that are grouped with that operation, but those grouped operations are not also predicated, the IR ends up inducing additional control flow and/or running unnecessary operations that may reduce performance, all while loosing the logical grouping of the change.

I think the basic MLIR traversals are lacking in this respect, but it's not hard to do generally. (it's a semi-recursive walker, I can make a separate PR if you like).

All that being said, if there's a strong preference for a predicated store here, I can make that change and we can argue the finer semantics in follow ups :D

Reply:
I do feel strongly that we should avoid inserting control flow in those passes unless it is strictly necessary. My main reason is that it blocks many things for the passes after, it prevents scheduling and other key optimizations at MLIR and LLVM IR.